horovod/common/ops/gpu_operations.h

```diff
@@ -14,70 +14,91 @@
 // limitations under the License.
 // =============================================================================
 
-#ifndef HOROVOD_CUDA_OPERATIONS_H
-#define HOROVOD_CUDA_OPERATIONS_H
+#ifndef HOROVOD_GPU_OPERATIONS_H
+#define HOROVOD_GPU_OPERATIONS_H
 
 #include <queue>
 #include <unordered_map>
 #include <vector>
 
+#if HAVE_CUDA
 #include <cuda_runtime.h>
+using gpuError_t = cudaError_t;
+using gpuEvent_t = cudaEvent_t;
+using gpuStream_t = cudaStream_t;
+#elif HAVE_ROCM
+#include <hip/hip_runtime_api.h>
+using gpuError_t = hipError_t;
+using gpuEvent_t = hipEvent_t;
+using gpuStream_t = hipStream_t;
+#endif
 
 #include "collective_operations.h"
 #include "../thread_pool.h"
 
 namespace horovod {
 namespace common {
 
-struct CUDAContext {
-  cudaError_t GetCudaEvent(cudaEvent_t* event);
-
-  cudaError_t ReleaseCudaEvent(cudaEvent_t event);
+class GPUContext {
+public:
+  GPUContext();
+  ~GPUContext();
```

Comment:
I ran into an issue desribed here https://herbsutter.com/gotw/_100/ .  Specifically: 

> You still need to write the visible class’ destructor yourself and define it out of line in the implementation file, even if normally it’s the same as what the compiler would generate. This is because although both unique_ptr and shared_ptr can be instantiated with an incomplete type, unique_ptr’s destructor requires a complete type in order to invoke delete (unlike shared_ptr which captures more information when it’s constructed). By writing it yourself in the implementation file, you force it to be defined in a place where impl is already defined, and this successfully prevents the compiler from trying to automatically generate the destructor on demand in the caller’s code where impl is not defined.

Reply:
Thanks @jeffdaily for the explanation and the link.